### PR TITLE
#1253: optimized url-updater

### PR DESCRIPTION
**Pull request by jan-vcapgemini**
Fixes/Addresses: #1253

Implements:

* added isMissingOs method to AbstractUrlUpdater (checks list of OS types for existence)
* added field of OS types to AbstractUrlUpdater
* added getters for URL and Checksum files to UrlVersion
* added check for existing checksum file to AbstractUrlUpdater checkDownloadUrl method
* added isMissingOs check to AbstractUrlUpdater update method
* added isMissingOs check to AndroidStudioUrlUpdater update method
* added isMissingOs check to IntellijUrlUpdater update method
* added new integration test for UrlUpdater

### CODE REVIEW

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java [hunk 1]
```diff
@@ -11,9 +11,7 @@
 import java.security.NoSuchAlgorithmException;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.Collection;
-import java.util.Locale;
-import java.util.Set;
+import java.util.*;
```
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:14]: You are using Intellij what has stupid default settings.
Star-imports are evil and discouraged but Intellij creates them by default.
You need to configure the number of imports per package required for star import to 999.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java [hunk 2]
```diff
@@ -437,6 +450,43 @@ private void doUpdateStatusJson(boolean success, int statusCode, UrlVersion urlV
...
+
+  /**
+   * Checks if an OS type was missing
+   *
+   * @param urlRepository {@link UrlRepository}
+   * @param version String of version folder
+   * @param osTypesOverwrite Set of OS types to overwrite
+   * @return true if an OS type was missing, false if not
+   */
+  public boolean isMissingOs(UrlRepository urlRepository, String version, Set<String> osTypesOverwrite) {
```
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:476]: why do we pass the Set of url-files to this method?
I would instead have a protected method `getUrlsFilenames()` that by default returns your constant `osTypes` (in Java constants should be named UPPER_CASE and strictly these are not Operating System Types but names of url-files).
You could also add a method `protected boolean isOsDependent()` that by default returns `true`. The implementation of `getUrlsFilenames()` can then be implemented as following:
```
if (isOsDependent()) {
  return URL_FILES_PER_OS;
} else {
  return URL_FILES_OS_INDEPENTENT;
}
```
Now you can implement like this:
```
`protected boolean isMissingOs(UrlVersion version) {
  for (String urlsFile : getUrlsFilenames()) {
    if (version.getChild(urlsFile) == null) {
       return true;
    }
  }
  return false;
}
```

Seems easier and now updaters that are OS independent can simply override `isOsDependent` and return `true`.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java [hunk 3]
```diff
@@ -458,7 +508,7 @@ public void update(UrlRepository urlRepository) {
         break;
       }
 
-      if (edition.getChild(version) == null) {
+      if (edition.getChild(version) == null || isMissingOs(urlRepository, version, null)) {
```
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:511]:
```
UrlVersion urlVersion = edition.getChild(version);
if ((urlVersion == null) || isMissingOs(urlVersion)) {
```

**Review by hohwille — CHANGES_REQUESTED** [review]
@jan-vcapgemini I had some pending review comments. Can you also have a look please?

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java (hunk 1 above)
Comment (jan-vcapgemini) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: Adjusted now.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java (hunk 3 above)
Comment (jan-vcapgemini) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: Adjusted, thanks.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java [hunk 4]
```diff
@@ -58,6 +60,13 @@ public abstract class AbstractUrlUpdater extends AbstractProcessorWithTimeout im
   /** {@link SystemArchitecture#ARM64}. */
   protected static final SystemArchitecture ARM64 = SystemArchitecture.ARM64;
 
+  /** List of URL file names dependent on OS which need to be checked for existence */
+  private static final Set<String> URL_FILENAMES_PER_OS = new HashSet<>(
+      Arrays.asList("linux_x64.urls", "mac_arm64.urls", "mac_x64.urls", "windows_x64.urls"));
+
+  /** List of URL file names independent of OS which need to be checked for existence */
+  private static final Set<String> URL_FILENAMES_OS_INDEPENDENT = new HashSet<>();
```
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:68]: should IMHO contain `urls` :)

**Review by hohwille — CHANGES_REQUESTED** [review]
@jan-vcapgemini almost perfect. Thanks 👍 I found one last problem that I commented. Please have a look.

**Review by hohwille — APPROVED** [review]
@jan-vcapgemini Excellent. Now ready to merge. Thanks 👍 
Nice-to-have improvement: Use `Set.of(...)` to make code more clean.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java (hunk 4 above)
Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: Also fixed now.

**Pull request merged**